### PR TITLE
Adding a logout request object

### PR DESCRIPTION
**Pull request by pkanher617**
This PR adds a logout request object to use during the logout call. Currently only used by msal-browser, may also be used by msal-node in the future.

### CODE REVIEW

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 1]
```diff
@@ -467,10 +468,11 @@ export class PublicClientApplication {
      * Default behaviour is to redirect the user to `window.location.href`.
      * @param logoutRequest 
      */
-    logout(account: AccountInfo, postLogoutRedirectUri?: string, authority?: string): void {
-        this.createAuthCodeClient(authority).then((authClient: AuthorizationCodeClient) => {
+    logout(logoutRequest: EndSessionRequest): void {
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts:471]: Thoughts on making this `async`? This seems like a good time, as you are changing the function signature.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 2]
```diff
@@ -96,12 +96,13 @@ export class AuthorizationCodeClient extends BaseClient {
...
      */
-    getLogoutUri(account: AccountInfo, postLogoutRedirectUri: string): string {
+    getLogoutUri(logoutRequest: EndSessionRequest): string {
         // Clear current account.
-        this.cacheManager.removeAccount(AccountEntity.generateAccountCacheKey(account));
+        this.cacheManager.removeAccount(AccountEntity.generateAccountCacheKey(logoutRequest.account));
 
         // Get postLogoutRedirectUri.
-        const postLogoutUriParam = postLogoutRedirectUri ? `?${AADServerParamKeys.POST_LOGOUT_URI}=` + encodeURIComponent(postLogoutRedirectUri) : "";
+        const postLogoutUriParam = logoutRequest.postLogoutRedirectUri ?
```
Comment (jasonnutter) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:104]: Can you we also add correlationId?

File: lib/msal-common/src/client/AuthorizationCodeClient.ts [hunk 3]
```diff
@@ -96,12 +96,13 @@ export class AuthorizationCodeClient extends BaseClient {
...
-    getLogoutUri(account: AccountInfo, postLogoutRedirectUri: string): string {
+    getLogoutUri(logoutRequest: EndSessionRequest): string {
         // Clear current account.
-        this.cacheManager.removeAccount(AccountEntity.generateAccountCacheKey(account));
+        this.cacheManager.removeAccount(AccountEntity.generateAccountCacheKey(logoutRequest.account));
 
         // Get postLogoutRedirectUri.
-        const postLogoutUriParam = postLogoutRedirectUri ? `?${AADServerParamKeys.POST_LOGOUT_URI}=` + encodeURIComponent(postLogoutRedirectUri) : "";
+        const postLogoutUriParam = logoutRequest.postLogoutRedirectUri ? 
+            `?${AADServerParamKeys.POST_LOGOUT_URI}=` + encodeURIComponent(logoutRequest.postLogoutRedirectUri) : "";
```
Comment (jasonnutter) [inline lib/msal-common/src/client/AuthorizationCodeClient.ts:105]:
```
?${AADServerParamKeys.POST_LOGOUT_URI}=${encodeURIComponent(logoutRequest.postLogoutRedirectUri)}`
```

File: lib/msal-common/src/request/EndSessionRequest.ts [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { AccountInfo } from "../account/AccountInfo";
+
+export type EndSessionRequest = {
+    account: AccountInfo,
+    postLogoutRedirectUri?: string,
+    authority?: string
```
Comment (jasonnutter) [inline lib/msal-common/src/request/EndSessionRequest.ts:11]: Correlation ID would be a good thing to add.

Comment (tnorling) [reply lib/msal-common/src/request/EndSessionRequest.ts:11]: Thoughts on also adding either id_token_hint or extraQueryParameters to support the ask in this issue #996 ?

Comment (pkanher617) [reply lib/msal-common/src/request/EndSessionRequest.ts:11]: I've added correlationId, we can discuss further changes to logout in a separate PR.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 1 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: done.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 2 above)
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts:115]: done.

File: lib/msal-common/src/client/AuthorizationCodeClient.ts (hunk 3 above)
Comment (pkanher617) [reply lib/msal-common/src/client/AuthorizationCodeClient.ts]: done.

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 5]
```diff
@@ -460,12 +461,12 @@ export class PublicClientApplication {
      * Default behaviour is to redirect the user to `window.location.href`.
      * @param logoutRequest 
      */
-    logout(account: AccountInfo, postLogoutRedirectUri?: string, authority?: string): void {
-        this.createAuthCodeClient(authority).then((authClient: AuthorizationCodeClient) => {
-            // create logout string and navigate user window to logout. Auth module will clear cache.
-            const logoutUri: string = authClient.getLogoutUri(account, postLogoutRedirectUri || this.getPostLogoutRedirectUri());
-            BrowserUtils.navigateWindow(logoutUri);
-        });
+    async logout(logoutRequest: EndSessionRequest): Promise<void> {
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts:464]: I believe `logoutRequest` can be marked as optional?

File: lib/msal-browser/src/app/PublicClientApplication.ts [hunk 6]
```diff
@@ -692,5 +693,12 @@ export class PublicClientApplication {
         return authCodeRequest;
     }
 
+    private generateLogoutRequest(logoutRequest: EndSessionRequest): EndSessionRequest {
```
Comment (jasonnutter) [inline lib/msal-browser/src/app/PublicClientApplication.ts:696]: `logoutRequest` optional.

File: lib/msal-browser/test/app/PublicClientApplication.spec.ts [hunk 7]
```diff
@@ -1061,19 +1061,17 @@ describe("PublicClientApplication.ts Class Unit Tests", () => {
...
                 expect(urlNavigate).to.be.eq(testLogoutUrl);
                 expect(noHistory).to.be.undefined;
                 done();
             });
-            const testAccount: AccountInfo = {
-                homeAccountId: TEST_DATA_CLIENT_INFO.TEST_HOME_ACCOUNT_ID,
-                environment: "login.windows.net",
-                tenantId: "testTenantId",
-                username: "test@contoso.com"
+            pca.logout({});
```
Comment (jasonnutter) [inline lib/msal-browser/test/app/PublicClientApplication.spec.ts:1070]: Should not have to pass an empty object.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 5 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: done.

File: lib/msal-browser/src/app/PublicClientApplication.ts (hunk 6 above)
Comment (pkanher617) [reply lib/msal-browser/src/app/PublicClientApplication.ts]: done.

File: lib/msal-browser/test/app/PublicClientApplication.spec.ts (hunk 7 above)
Comment (pkanher617) [reply lib/msal-browser/test/app/PublicClientApplication.spec.ts]: fixed.

**Comment by stephenstroud** [comment]
Is this PR due out? I have a requirement where ID_TOKEN_HINT needs to be passed in for MSAL v1? Is this PR open to contribution?

**Comment by tnorling** [comment]
@stephenstroud This PR was released as part of version 2.0.0 but does not include functionality for providing `id_token_hint`. You can track issue #996 for that feature request, but it will likely not make it into version 1 at this point as new features are going into version 2.x moving forward.

**Pull request merged**